astropy/utils/tests/test_data.py

```diff
@@ -1881,96 +1937,82 @@ def test_download_parallel_respects_pkgname(temp_cache, valid_urls):
     assert len(get_cached_urls(pkgname=a)) == FEW
 
 
-@pytest.mark.parametrize("b", _shelve_possible_backends)
-def test_cache_with_different_shelve_backends(b, temp_cache, valid_urls):
-    """Without special handling this emits a warning for dbm.dumb."""
-    create_cache_with_backend(b)
+def test_removal_of_open_files(temp_cache, valid_urls):
+    if not can_rename_directory_in_use():
+        pytest.skip("This platform is not able to remove files while in use.")
+    u, c = next(valid_urls)
+    with open(download_file(u, cache=True)):
+        clear_download_cache(u)
+        assert not is_url_in_cache(u)
+        check_download_cache()
+    check_download_cache()
 
-    uc = list(islice(valid_urls, FEW))
-    for u, c in uc:
-        download_file(u, cache=True)
-        assert is_url_in_cache(u)
 
-    for u, _ in uc:
+def test_update_of_open_files(temp_cache, valid_urls):
+    if not can_rename_directory_in_use():
+        pytest.skip("This platform is not able to remove files while in use.")
+    u, c = next(valid_urls)
+    with open(download_file(u, cache=True)):
+        u2, c2 = next(valid_urls)
+        f = download_file(u, cache='update', sources=[u2])
+        check_download_cache()
         assert is_url_in_cache(u)
-
+        assert get_file_contents(f) == c2
+    assert is_url_in_cache(u)
     check_download_cache()
 
-    for u, _ in uc:
-        assert is_url_in_cache(u)
 
-    for u, c in uc:
-        assert get_file_contents(
-            download_file(u, cache=True, sources=[])) == c
+def test_removal_of_open_files_windows(temp_cache, valid_urls, monkeypatch):
+    def no_rmtree(*args, **kwargs):
+        warnings.warn(CacheMissingWarning("in use", ""))
+        raise PermissionError
+    if can_rename_directory_in_use():
+        # This platform is able to remove files while in use.
+        monkeypatch.setattr(astropy.utils.data, "_rmtree", no_rmtree)
 
-    clear_download_cache()
+    u, c = next(valid_urls)
+    with open(download_file(u, cache=True)):
+        with catch_warnings(CacheMissingWarning) as ws:
+            clear_download_cache(u)
+        in_use = False
+        for w in ws:
+            if "in use" in str(w.message):
+                in_use = True
+        assert in_use, "No warning mentions that a file is in use"
```

Comment:
I think this is what you are trying to do?
```suggestion
        with pytest.warns(CacheMissingWarning, match=r".* in use.*"):
            clear_download_cache(u)
```

Reply:
yes thanks! all `catch_warnings` now obliterated

Reply:
Yeah, sorry about the conflict. It is caused by my work at #10504 to remove the `catch_warnings` as well.